modules/fedimint-lnv2-client/src/lib.rs

```diff
@@ -0,0 +1,725 @@
+mod receive_sm;
+mod send_sm;
+
+use std::collections::BTreeMap;
+use std::sync::Arc;
+
+use anyhow::{anyhow, ensure, Context};
+use async_stream::stream;
+use bitcoin_hashes::{sha256, Hash};
+use fedimint_client::module::init::{ClientModuleInit, ClientModuleInitArgs};
+use fedimint_client::module::recovery::NoModuleBackup;
+use fedimint_client::module::{ClientContext, ClientModule, IClientModule};
+use fedimint_client::oplog::UpdateStreamOrOutcome;
+use fedimint_client::sm::util::MapStateTransitions;
+use fedimint_client::sm::{DynState, ModuleNotifier, State, StateTransition};
+use fedimint_client::transaction::{ClientOutput, TransactionBuilder};
+use fedimint_client::{sm_enum_variant_translation, DynGlobalClientContext};
+use fedimint_core::api::DynModuleApi;
+use fedimint_core::config::FederationId;
+use fedimint_core::core::{IntoDynInstance, ModuleInstanceId, OperationId};
+use fedimint_core::db::{DatabaseTransaction, DatabaseVersion};
+use fedimint_core::encoding::{Decodable, Encodable};
+use fedimint_core::module::{
+    ApiVersion, CommonModuleInit, ModuleCommon, ModuleInit, MultiApiVersion, TransactionItemAmount,
+};
+use fedimint_core::time::duration_since_epoch;
+use fedimint_core::util::SafeUrl;
+use fedimint_core::{apply, async_trait_maybe_send, Amount};
+use fedimint_lnv2_common::api::LnFederationApi;
+use fedimint_lnv2_common::config::LightningClientConfig;
+use fedimint_lnv2_common::contracts::{IncomingContract, OutgoingContract};
+use fedimint_lnv2_common::{
+    LightningClientContext, LightningCommonInit, LightningModuleTypes, LightningOutput,
+};
+use futures::StreamExt;
+use lightning_invoice::Bolt11Invoice;
+use secp256k1::{ecdh, KeyPair, PublicKey, Scalar, SecretKey};
+use serde::{Deserialize, Serialize};
+use tpe::{derive_agg_decryption_key, AggregateDecryptionKey};
+
+use crate::receive_sm::{Pending, ReceiveSMState, ReceiveStateMachine};
+use crate::send_sm::{SendSMState, SendStateMachine};
+
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub struct LightningOperationMeta;
+
+/// Number of blocks until outgoing lightning contracts time out and user
+/// client can refund it unilaterally
+const SEND_EXPIRATION_DELTA_BLOCKS_DEFAULT: u64 = 500;
+
+/// Default expiration time for lightning invoices
+const INVOICE_EXPIRATION_SECONDS_DEFAULT: u32 = 3600;
+
+/// The high-level state of an payment operation over lightning
+#[derive(Debug, Clone, Eq, PartialEq, Serialize, Deserialize)]
+#[serde(rename_all = "snake_case")]
+pub enum SendState {
+    Funding,
+    Funded,
```

Comment:
I think one of these states should emit the `block_height` at which the outgoing contract is redeemable. I believe wallets use that today to show in the UI